### PR TITLE
Feature/smooth ndt map update

### PR DESCRIPTION
**Pull request by anhnv3991**
- Add copy and move constructors/assignment operators to NDT and voxel grid covariance

### CODE REVIEW

**Comment by SakodaShintaro** [comment]
Thank you for the pull request!
The current default branch for this repository is `tier4/main`, so I changed the pull request destination from `master` to `tier4/main`.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp [hunk 1]
```diff
@@ -56,6 +56,77 @@
 #include <pcl/filters/boost.h>
 #include "multigrid_pclomp/multi_voxel_grid_covariance_omp.h"
 
+template <typename PointT>
+pclomp::MultiVoxelGridCovariance<PointT>::MultiVoxelGridCovariance(const MultiVoxelGridCovariance& other)
+: pcl::VoxelGrid<PointT>(other),
+  leaves_(other.leaves_),
+  grid_leaves_(other.grid_leaves_),
+  leaf_indices_(other.leaf_indices_),
+  kdtree_(other.kdtree_)
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp]: One my concern is that the `kdtree_` has some `shared_ptr`s. Is this copy shallow copy?
https://github.com/PointCloudLibrary/pcl/blob/1033d24b750eed4f22c9c1c7711398aba7877055/kdtree/include/pcl/kdtree/kdtree_flann.h#L288-L294

Comment (SakodaShintaro) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp]: I will make a note of what anhnv3991-san told me.

(1) `flann_index_` is reset each time setInputCloud is used, so the target will change at that time.
https://github.com/PointCloudLibrary/pcl/blob/1033d24b750eed4f22c9c1c7711398aba7877055/kdtree/include/pcl/kdtree/impl/kdtree_flann.hpp#L131-L135

(2) `cloud_` is reset with convertCloudToArray. convertCloudToArray is also called in setInputCloud.
https://github.com/PointCloudLibrary/pcl/blob/1033d24b750eed4f22c9c1c7711398aba7877055/kdtree/include/pcl/kdtree/impl/kdtree_flann.hpp#L428-L498

Therefore, it is considered that there is no practical problem with shallow copying.

File: include/multigrid_pclomp/multigrid_ndt_omp.h [hunk 2]
```diff
@@ -183,6 +214,11 @@ class MultiGridNormalDistributionsTransform : public pcl::Registration<PointSour
     return (resolution_);
   }
 
+  inline Eigen::Vector3f getGridLeafSize() const 
+  {
+    return target_cells_.getLeafSize();
+  }
+
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multigrid_ndt_omp.h]: Why was `getGridLeafSize` added?
I searched "getGridLeafSize" in autoware and it didn't seem to be used anywhere.

File: include/multigrid_pclomp/multigrid_ndt_omp.h [hunk 3]
```diff
@@ -318,7 +354,7 @@ class MultiGridNormalDistributionsTransform : public pcl::Registration<PointSour
   using pcl::Registration<PointSource, PointTarget>::reg_name_;
   using pcl::Registration<PointSource, PointTarget>::getClassName;
   using pcl::Registration<PointSource, PointTarget>::input_;
-  using pcl::Registration<PointSource, PointTarget>::indices_;
+  using pcl::Registration<PointSource, PointTarget>::indices_;  // Why need this one?
```
Comment (SakodaShintaro) [inline include/multigrid_pclomp/multigrid_ndt_omp.h]: I also don't think this declaration is necessary.
Similarly, the following four lines near here seem unnecessary.
````
   using pcl::Registration<PointSource, PointTarget>::getClassName;
...
   using pcl::Registration<PointSource, PointTarget>::indices_; // Why need this one?
...
   using pcl::Registration<PointSource, PointTarget>::corr_dist_threshold_;
   using pcl::Registration<PointSource, PointTarget>::inlier_threshold_;
````
I think it's a good idea to delete it.

**Review by SakodaShintaro — COMMENTED** [review]
Added two small comments

File: include/multigrid_pclomp/multigrid_ndt_omp.h (hunk 2 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multigrid_ndt_omp.h]: @SakodaShintaro Deleted it.

File: include/multigrid_pclomp/multigrid_ndt_omp.h (hunk 3 above)
Comment (anhnv3991) [reply include/multigrid_pclomp/multigrid_ndt_omp.h]: @SakodaShintaro I deleted these linese too.

**Comment by anhnv3991** [comment]
@SakodaShintaro I just rebase the tier4/main branch, so could you please check again?

**Review by SakodaShintaro — APPROVED** [review]
I don't know why, but once the auto format is committed in pre-commit, the build check results are not visible in this pull request.
The build check itself is being executed, and  everything is OK.
https://github.com/tier4/ndt_omp/actions/runs/7970490513

I also checked the operation of logging_simulator again using Autoware and it worked properly. :100:

**Comment by anhnv3991** [comment]
@SakodaShintaro Thanks!

**Pull request merged**